File: vps/bootstrap.ts
#!/usr/bin/env -S npx tsx
import { $ } from "dax";
import {
  chdirFromImportMeta,
  compileLinuxBinary,
  requireVpsHost,
  runRemoteBinaryAndCleanup,
} from "./utils.ts";

chdirFromImportMeta(import.meta.url, "..");

const VPS_HOST = requireVpsHost("bootstrap");

await $`mkdir -p vps/dist`;
await compileLinuxBinary("vps/bootstrap-remote.ts", "vps/dist/bootstrap-remote");

await $`ssh ubuntu@${VPS_HOST} 'sudo mkdir -p /opt/bootstrap/vps'`;
await $`rsync -a --delete vps/ ubuntu@${VPS_HOST}:/tmp/bootstrap-vps/`;
await $`ssh ubuntu@${VPS_HOST} 'sudo rsync -a --delete /tmp/bootstrap-vps/ /opt/bootstrap/vps/'`;

await $`rsync vps/dist/bootstrap-remote ubuntu@${VPS_HOST}:/tmp/bootstrap-remote`;
await runRemoteBinaryAndCleanup({
  host: VPS_HOST,
  user: "ubuntu",
  remotePath: "/tmp/bootstrap-remote",
  sudo: true,
});

File: minibuild/lib/projects/bootstrap.ml
open Shared

let ssh_host host = Printf.sprintf "ubuntu@%s" host

let require_vps_host () =
  match Sys.getenv_opt "VPS_HOST" with
  | Some host when String.trim host <> "" -> Ok (String.trim host)
  | _ ->
    Error
      (Types.String_error
         "VPS_HOST is required (set it in mise env or shell env before running bootstrap)")
;;

let run =
  Step.create
    ~name:"bootstrap vps"
    ~apply:(fun ctx ->
      let* host = require_vps_host () in
      let target = ssh_host host in
      let repo_root = Eio.Path.native_exn ctx.project_root in
      let vps_dir = Filename.concat repo_root "vps/" in
      let* _ =
        Cmd.remote_cmd ~ctx target "sudo mkdir -p /opt/bootstrap/vps"
        |> Result.map_error (fun e -> Types.CMD_Error e)
      in
      let* _ =
        Cmd.run_cmd
          ~ctx
          [ "rsync"
          ; "-a"
          ; "--delete"
          ; vps_dir
          ; Printf.sprintf "%s:/tmp/bootstrap-vps/" target
          ]
        |> Result.map_error (fun e -> Types.CMD_Error e)
      in
      let remote_cmd =
        "sudo rsync -a --delete /tmp/bootstrap-vps/ /opt/bootstrap/vps/ && sudo bash /opt/bootstrap/vps/bootstrap.sh"
      in
      let* _ =
        Cmd.remote_cmd ~ctx target remote_cmd
        |> Result.map_error (fun e -> Types.CMD_Error e)
      in
      Ok ctx)
    ()
;;
